File: tests/engine/integration.test.ts
import { describe, it, expect } from "vitest";
import { generateWords } from "../../src/engine/word-generator.js";
import {
  createInputState,
  processChar,
  processSpace,
  processBackspace,
  processCtrlBackspace,
} from "../../src/engine/input-processor.js";
import {
  calculateStats,
  calculateConsistency,
} from "../../src/engine/stats-calculator.js";
import {
  createTimer,
  startTimer,
  getElapsedSeconds,
  getRemainingSeconds,
} from "../../src/engine/timer.js";

describe("full typing test flow", () => {
  it("simulates a complete word-mode test", () => {
    // 1. Generate words
    const words = generateWords({
      language: "english",
      count: 3,
      punctuation: false,
      numbers: false,
    });
    expect(words).toHaveLength(3);

    // 2. Create input state
    let state = createInputState(words);
    expect(state.finished).toBe(false);

    // 3. Type each word perfectly + space
    for (let w = 0; w < words.length; w++) {
      for (const ch of words[w]) {
        state = processChar(state, ch);
      }
      state = processSpace(state);
    }

    // 4. Should be finished
    expect(state.finished).toBe(true);

    // 5. Calculate stats
    const stats = calculateStats(state, 10);
    expect(stats.wpm).toBeGreaterThan(0);
    expect(stats.accuracy).toBe(100);
    expect(stats.incorrectChars).toBe(0);
    expect(stats.extraChars).toBe(0);
    expect(stats.missedChars).toBe(0);
  });

  it("simulates a test with errors and corrections", () => {
    const words = ["hello", "world"];
    let state = createInputState(words);

    // Type "hello" with a mistake, then fix it
    state = processChar(state, "h");
    state = processChar(state, "x"); // typo
    state = processBackspace(state);  // fix it
    state = processChar(state, "e");
    state = processChar(state, "l");
    state = processChar(state, "l");
    state = processChar(state, "o");
    state = processSpace(state);

    // Type "world" correctly
    for (const ch of "world") {
      state = processChar(state, ch);
    }
    state = processSpace(state);

    expect(state.finished).toBe(true);

    const stats = calculateStats(state, 30);
    // Both words typed correctly (after correction)
    expect(stats.wpm).toBeGreaterThan(0);
    // Accuracy should be less than 100 because of the corrected mistake
    expect(stats.accuracy).toBeLessThan(100);
    expect(stats.accuracy).toBeGreaterThan(0);
  });

  it("simulates ctrl+backspace during typing", () => {
    let state = createInputState(["test", "word"]);

    // Type half a word, then clear it
    state = processChar(state, "t");
    state = processChar(state, "x");
    state = processChar(state, "y");
    state = processCtrlBackspace(state);

    expect(state.words[0].typed).toBe("");
    expect(state.cursorPosition).toBe(0);
    expect(state.currentWordIndex).toBe(0);

    // Now type it correctly
    for (const ch of "test") {
      state = processChar(state, ch);
    }
    state = processSpace(state);
    expect(state.currentWordIndex).toBe(1);
  });

  it("calculates consistency from WPM history", () => {
    // Simulate a steady typer
    const steadyHistory = [60, 62, 58, 61, 59, 60];
    const steadyConsistency = calculateConsistency(steadyHistory);
    expect(steadyConsistency).toBeGreaterThan(90);

    // Simulate an erratic typer
    const erraticHistory = [20, 80, 15, 90, 25, 75];
    const erraticConsistency = calculateConsistency(erraticHistory);
    expect(erraticConsistency).toBeLessThan(40);

    // Steady should beat erratic
    expect(steadyConsistency).toBeGreaterThan(erraticConsistency);
  });

  it("timer works with the full test flow", () => {
    // Create and start a 30-second timer
    const timer = createTimer(30);
    const started = startTimer(timer);

    expect(started.running).toBe(true);
    expect(getElapsedSeconds(started)).toBe(0);
    expect(getRemainingSeconds(started)).toBe(30);

    // Simulate some elapsed time
    const midway = { ...started, elapsedMs: 15000 };
    expect(getElapsedSeconds(midway)).toBe(15);
    expect(getRemainingSeconds(midway)).toBe(15);
  });

  it("handles extra characters in stats", () => {
    let state = createInputState(["hi"]);
    state = processChar(state, "h");
    state = processChar(state, "i");
    state = processChar(state, "x"); // extra
    state = processChar(state, "y"); // extra
    state = processSpace(state);

    const stats = calculateStats(state, 10);
    expect(stats.extraChars).toBe(2);
    // Word has extra chars so it's not counted as correct for WPM
    expect(stats.wpm).toBe(0);
    expect(stats.rawWpm).toBeGreaterThan(0);
  });

  it("handles missed characters in stats", () => {
    let state = createInputState(["hello", "ok"]);
    // Type only "he" then skip to next word
    state = processChar(state, "h");
    state = processChar(state, "e");
    state = processSpace(state);

    const stats = calculateStats(state, 10);
    expect(stats.missedChars).toBe(3); // l, l, o
  });

  it("going back to previous word preserves state", () => {
    let state = createInputState(["hello", "world"]);

    // Type "hello" and advance
    for (const ch of "hello") {
      state = processChar(state, ch);
    }
    state = processSpace(state);
    expect(state.currentWordIndex).toBe(1);

    // Go back to "hello"
    state = processBackspace(state);
    expect(state.currentWordIndex).toBe(0);
    expect(state.words[0].typed).toBe("hello");
    expect(state.words[0].completed).toBe(false);
    expect(state.cursorPosition).toBe(5);
  });
});

File: tests/integration_test.rs
use buddytype::engine::input_processor::*;
use buddytype::engine::stats_calculator::*;
use buddytype::engine::timer::*;
use buddytype::engine::word_generator::generate_words;
use buddytype::types::GeneratorConfig;
use std::time::{Duration, Instant};

#[test]
fn simulates_complete_word_mode_test() {
    // 1. Generate words
    let words = generate_words(&GeneratorConfig {
        language: "english".into(),
        count: 3,
        punctuation: false,
        numbers: false,
    });
    assert_eq!(words.len(), 3);

    // 2. Create input state
    let mut state = create_input_state(&words);
    assert!(!state.finished);

    // 3. Type each word perfectly + space
    for w in &words {
        for ch in w.chars() {
            state = process_char(state, ch);
        }
        state = process_space(state);
    }

    // 4. Should be finished
    assert!(state.finished);

    // 5. Calculate stats
    let stats = calculate_stats(&state, 10.0);
    assert!(stats.wpm > 0.0);
    assert_eq!(stats.accuracy, 100.0);
    assert_eq!(stats.incorrect_chars, 0);
    assert_eq!(stats.extra_chars, 0);
    assert_eq!(stats.missed_chars, 0);
}

#[test]
fn simulates_test_with_errors_and_corrections() {
    let words: Vec<String> = vec!["hello".into(), "world".into()];
    let mut state = create_input_state(&words);

    // Type "hello" with a mistake, then fix it
    state = process_char(state, 'h');
    state = process_char(state, 'x'); // typo
    state = process_backspace(state);
    state = process_char(state, 'e');
    state = process_char(state, 'l');
    state = process_char(state, 'l');
    state = process_char(state, 'o');
    state = process_space(state);

    // Type "world" correctly
    for ch in "world".chars() {
        state = process_char(state, ch);
    }
    state = process_space(state);

    assert!(state.finished);

    let stats = calculate_stats(&state, 30.0);
    assert!(stats.wpm > 0.0);
    assert!(stats.accuracy < 100.0);
    assert!(stats.accuracy > 0.0);
}

#[test]
fn simulates_ctrl_backspace_during_typing() {
    let mut state = create_input_state(&["test".into(), "word".into()]);

    state = process_char(state, 't');
    state = process_char(state, 'x');
    state = process_char(state, 'y');
    state = process_ctrl_backspace(state);

    assert_eq!(state.words[0].typed, "");
    assert_eq!(state.cursor_position, 0);
    assert_eq!(state.current_word_index, 0);

    for ch in "test".chars() {
        state = process_char(state, ch);
    }
    state = process_space(state);
    assert_eq!(state.current_word_index, 1);
}

#[test]
fn calculates_consistency_from_wpm_history() {
    let steady = calculate_consistency(&[60.0, 62.0, 58.0, 61.0, 59.0, 60.0]);
    assert!(steady > 90.0);

    let erratic = calculate_consistency(&[20.0, 80.0, 15.0, 90.0, 25.0, 75.0]);
    assert!(erratic < 40.0);

    assert!(steady > erratic);
}

#[test]
fn timer_works_with_full_test_flow() {
    let timer = create_timer(Some(30));
    let now = Instant::now();
    let started = start_timer(timer, now);

    assert!(started.running);
    assert_eq!(get_elapsed_seconds(&started), 0.0);
    assert_eq!(get_remaining_seconds(&started), 30.0);

    // Simulate some elapsed time
    let later = now + Duration::from_secs(15);
    let midway = tick_timer(started, later);
    assert!((get_elapsed_seconds(&midway) - 15.0).abs() < 0.01);
    assert!((get_remaining_seconds(&midway) - 15.0).abs() < 0.01);
}

#[test]
fn handles_extra_characters_in_stats() {
    let mut state = create_input_state(&["hi".into()]);
    state = process_char(state, 'h');
    state = process_char(state, 'i');
    state = process_char(state, 'x'); // extra
    state = process_char(state, 'y'); // extra
    state = process_space(state);

    let stats = calculate_stats(&state, 10.0);
    assert_eq!(stats.extra_chars, 2);
    assert_eq!(stats.wpm, 0.0); // extra chars means word not correct
    assert!(stats.raw_wpm > 0.0);
}

#[test]
fn handles_missed_characters_in_stats() {
    let mut state = create_input_state(&["hello".into(), "ok".into()]);
    state = process_char(state, 'h');
    state = process_char(state, 'e');
    state = process_space(state);

    let stats = calculate_stats(&state, 10.0);
    assert_eq!(stats.missed_chars, 3); // l, l, o
}

#[test]
fn going_back_to_previous_word_preserves_state() {
    let mut state = create_input_state(&["hello".into(), "world".into()]);

    for ch in "hello".chars() {
        state = process_char(state, ch);
    }
    state = process_space(state);
    assert_eq!(state.current_word_index, 1);

    state = process_backspace(state);
    assert_eq!(state.current_word_index, 0);
    assert_eq!(state.words[0].typed, "hello");
    assert!(!state.words[0].completed);
    assert_eq!(state.cursor_position, 5);
}
